pygmt/clib/session.py

```diff
@@ -791,11 +791,61 @@ def put_vector(self, dataset, column, vector):
                 )
             )
 
+    def put_strings(self, dataset, family, strings):
+        """
+        Attach a numpy 1D array of dtype str as a column on a GMT dataset.
+
+        Use this function to attach string type numpy array data to a GMT
+        dataset and pass it to GMT modules. Wraps ``GMT_Put_Strings``.
+
+        The dataset must be created by :meth:`~gmt.clib.Session.create_data`
+        first.
+
+        .. warning::
+            The numpy array must be C contiguous in memory. If it comes from a
+            column slice of a 2d array, for example, you will have to make a
+            copy. Use :func:`numpy.ascontiguousarray` to make sure your vector
+            is contiguous (it won't copy if it already is).
+
+        Parameters
+        ----------
+        dataset : :class:`ctypes.c_void_p`
+            The ctypes void pointer to a ``GMT_Dataset``. Create it with
+            :meth:`~gmt.clib.Session.create_data`.
+        family : str
+            The family type of the dataset. Can be either ``GMT_IS_VECTOR`` or
+            ``GMT_IS_MATRIX``.
+        strings : numpy 1d-array
+            The array that will be attached to the dataset. Must be a 1d C
+            contiguous array.
+
+        Raises
+        ------
+        GMTCLibError
+            If given invalid input or ``GMT_Put_Strings`` exits with status !=
+            0.
+
+        """
+        c_put_strings = self.get_libgmt_func(
+            "GMT_Put_Strings",
+            argtypes=[ctp.c_void_p, ctp.c_uint, ctp.c_void_p, ctp.c_void_p],
+            restype=ctp.c_int,
+        )
+
+        strings_pointer = strings.ctypes.data_as(ctp.c_char_p)
+        status = c_put_strings(
+            self.session_pointer, self[family], dataset, strings_pointer
+        )
```

Comment:
> The following Python code can correctly pass strings array to the myfunc function in test.so library:
> 
> ```python
> # ...
> arr = (ctypes.c_char_p * len(string))()
> arr[:] = [s.encode() for s in string]
> lib.myfunc(arr)
> ```
> 
> The Python trick comes from https://stackoverflow.com/questions/3494598/. It works but I don't understand the details, and don't know if it's the simplest way.

Yes I tried this following your datetime example, but I think (?) it's functionally equivalent to [`strings.ctypes.data_as(ctp.c_char_p)`](https://numpy.org/doc/stable/reference/generated/numpy.ndarray.ctypes.html)? Either way it runs, but the string data doesn't seem to get written to the virtualfile so the test fails. I've also tried `strings.ctypes.data_as(ctp.POINTER(ctp.c_char_p))` but it doesn't seem to make a difference. :confused:

Reply:
`strings.ctypes.data_as(ctp.c_char_p)` may not work. 

What I did is adding a print statement inside the GMT_Put_Strings function, printing the string arrays passed to it. Using `strings.ctypes.data_as(ctp.c_char_p)` crashes immediately for me, but using my code can print the strings correctly, but then crash 🤦‍♂️.

Reply:
How about using:

```python
c_put_strings = self.get_libgmt_func(
    "GMT_Put_Strings",
    argtypes=[
        ctp.c_void_p,
        ctp.c_uint,
        ctp.c_void_p,
        ctp.POINTER(ctp.c_char_p),
    ],
    restype=ctp.c_int,
)
strings_pointer = (ctp.c_char_p * len(strings))()
strings_pointer[:] = np.char.encode(strings)
```

Does it crash for you?

Reply:
Yes, it still crashes, but I can see that the strings are correctly passed to the GMT_Put_Strings function.

Reply:
What about `strings.ctypes.data_as(ctp.POINTER(ctp.c_char_p))`? Are the strings correctly passed in?

Reply:
No, just segmentation fault

Reply:
After a little debugging on the GMT C code, it looks a potential GMT bug.

Reply:
See https://github.com/GenericMappingTools/gmt/pull/3718 for the bug report.

Reply:
Yep, I was just taking a look at it (thanks for tracking down the bug by the way!). Sounds like we'll need to pin to GMT > 6.1.1 for the next release?

Reply:
Actually I'm not 100% sure https://github.com/GenericMappingTools/gmt/pull/3718 is related to the issue here. I just had the feeling that the string array may be freed by Python when GMT tries to read it. 

Hopefully, GMT 6.1.1 can fix issue and #515.

Reply:
FYI, https://github.com/GenericMappingTools/gmt/pull/3718 is already merged into GMT master, and will be backported to 6.1 branch soon.

Reply:
Great! I'll test it out when I get the time (need to prepare some stuff for an online conference next week). Still need to wait on the grid problem at #515 but that's a separate issue.

Reply:
Just triggered the GMT master branch testing. Ideally, we should only see one failure from `test_put_strings`.